Scrape props

Sarahtonein opened the pull request:
**Whats New**
- Scrape proposals (unlimited message count, up to a 10k char limit to align with Snapshot proposal size limits)
- Concat and format messages accordingly
- Warn the user with Ephemeral messages if a proposal fails to be published (Title too long, Proposal too long, Missing Voting Options, etc)
- Ability to build on this, to allow for different voting options

**What's changed**
- Removed proposal modal
- Removed selects
- Removed buttons
- Updated README
- Ordered tasks from most frequent => least frequent
- Changed how we pass to Snapshot for the proposal

**How to test**
- Setup bot locally and use development server
- Create thread 
- Copy link / thread ID
- invoke /create_proposal <thread ID/Link>
- Vote for proposal and observe results

**Tested**
Create a proposal with various parameters broken (title too long, proposal too long, voting options not included)

Example Proposal from these changes: https://testnet.snapshot.box/#/s-tn:testnet-1.eth/proposal/0xd887d80065fa652b3b9d106bf5aaf831cba3ff6ffaf41c47f1e47133ab38315e

**Notes**
- Bot commands may need to be updated (.update_commands) once bot has been redeployed. If you are using Discord web-app, you may need to clear your cache.

bagelface commented on README.md:
````diff
@@ -181,6 +152,8 @@ Reassess: 0
 Abstain: 0
 ```
 
+Upon passing the Discord phase, the vote will then be published to Snapshot. A 72 hours voting phase begins. At this point the bot has successfully passed off the handling of the proposal to Snapshot and no longer tracks it.
````
"A 72 [hour] voting phase begins"

Sarahtonein replied:
Amended

bagelface commented on README.md:
````diff
@@ -98,79 +98,50 @@ You can type ```/help``` to get details about what commands can be used, along w
+Start drafting your proposal in the #drafting-props channel and select the appropriate tag (budget, or governance). Depending on your tag selected the bot will handle the proposal prefix. Just create a name for your proposal.
 
-If you wish to change the emojis used when drafting & reaching soft consensus prior to publishing to snapshot this can be done in consts/constants.py. Simply replace YES_VOTE / NO_VOTE or ABSTAIN_VOTE with the desired emoji id. For example: 
+Please consider and use the following stuctures for the proposals body, for uniformity and structural reasons when making a proposal draft:
````
"When making a draft, please use the following format for the proposal's body to ensure consistency across proposals:"

bagelface commented on README.md:
````diff
@@ -98,79 +98,50 @@ You can type ```/help``` to get details about what commands can be used, along w
-```
-YES_VOTE = "<:gunta:1199583728129802322>"
-```
+**Governance (improvement-props):**
````
why is this stylized like this? Maybe just do something like:
"Governance (general, non-budgetary proposals)"

Sarahtonein replied:
If you refer to my prior comment it answers this one :) 

> I was including the channel name for the type of proposal, I can remove the bracket text entirely if that clears it up, otherwise I think we could scrap this entire section and link the users to Snapshot proposals BGP / BBP #0 for details on constructing proposals.
> 
> https://snapshot.org/#/s:gov.bloomstudio.eth/proposal/0x885326576ef6829d4ceefe00cb69a60ab1dceac8cc1cc50a2eef53a7ca2332a8
> 
> https://snapshot.org/#/s:gov.bloomstudio.eth/proposal/0xfeacc5218c02c2b2d8d7a43f89e81340c88b1b4ac906a1369833bb1a60b6c14f

bagelface commented on README.md:
````diff
@@ -98,79 +98,50 @@ You can type ```/help``` to get details about what commands can be used, along w
-```
-Topic/Vote 1: new title
+- Voting Options (required)
+Anything can be posted here, just include the words 'voting options' for now. The bot will use default settings (Adopt, Reassess, Abstain) which will be dynamcially adjustable in the future
````
Sort of confusing for this to be marked as "required" but then also explain that you just need to include a placeholder value. Can this section be left out entirely and just handled by the bot without requiring specific user input?

bagelface commented on proposals/proposals.py:
```diff
@@ -109,53 +59,92 @@ async def prepare_draft(
+            logger.error(f"Guild not found: {guild_id}")
+            return None
+
+        channel_name = (
```
might be better to throw an error if there is not channel match, rather than just an ternary that falls back to governance channel

bagelface commented on proposals/thread_parser.py:
```diff
@@ -0,0 +1,159 @@
+    def extract_thread_id(thread_input: str) -> int:
+        """Extract thread ID from either a URL or direct ID"""
+        try:
+            if "discord.com" in thread_input:
```
seems like it should check for existence of `/threads/` in the url as well

Sarahtonein replied:
I don't see /threads in urls when copying specific threads or channels the format always seems to be discord.com/channel/thread, both being a string of numbers if there is any.

bagelface replied:
your urls aren't formatted like `https://discord.com/channels/1152979839737925705/1191283932096905307/threads/1268299037212414065`?

Sarahtonein replied:
> your urls aren't formatted like `https://discord.com/channels/1152979839737925705/1191283932096905307/threads/1268299037212414065`?

Nope! I disabled developer mode, and then they started formatting like that. I'll add support for both.

bagelface commented on proposals/thread_parser.py:
```diff
@@ -0,0 +1,159 @@
+                        "Your proposal content exceeds Snapshot's 10,000 character limit. Please shorten your proposal and try again.",
+                    )
+
+            if not found_voting_options:
```
why not just insert the voting options if they're not found?

Sarahtonein replied:
Makes sense to do so. I initially had this so that the bot knew when to stop collecting messages in a thread in the scrape so that it doesn't include miscellaneous messages from the author, or other commentators (as we filter them when scraping threads)

bagelface reviewed (changes requested):
a few small changes but otherwise looks good

Sarahtonein commented on README.md:
````diff
@@ -98,79 +98,50 @@ You can type ```/help``` to get details about what commands can be used, along w
+Creating a proposal requires careful consideration. Once you have launched a proposal it cannot be modified for transparency reasons. 
 
-When you decide to publish the draft, you can request the bot to do this.
+Start drafting your proposal in the #drafting-props channel and select the appropriate tag (budget, or governance). Depending on your tag selected the bot will handle the proposal prefix. Just create a name for your proposal.
````
Amended language here, and added a hyperlink for the channel.

bagelface commented on README.md:
````diff
@@ -98,79 +98,50 @@ You can type ```/help``` to get details about what commands can be used, along w
-"Edit" will allow you to select an an existing proposal to edit.
-"Preview" will allow you to preview an existing proposal before it is published.
-"Delete" will allow you delete and existing proposal that has not been published.
+**Budget (budgetary-props)**
````
ooh i see, it wasn't clear to me they were channels. Maybe make them links or add a # to beginning so it's more clear

bagelface commented on proposals/thread_parser.py:
```diff
@@ -0,0 +1,171 @@
+            return {
+                "title": thread.name,
+                "type": "budget"
+                if "budget" in thread.applied_tags[0].name.lower()
```
can this be assigned a variable and used here and in the logic block above

Sarahtonein replied:
Done :)

Sarahtonein commented:
> one last change and then it looks good to me

Modified as requested.

As a FYI I noticed that the Discord usernames if published as a @username were displaying the UID when publishing to Snapshot, added a func to convert these to the username.

Merged in Quorum Calc as this was approved.

bagelface commented on proposals/thread_parser.py:
```diff
@@ -0,0 +1,188 @@
+            governance_tag = False
+            budget_tag = False
+
+            if thread.applied_tags:
```
this should probably throw an explicit error if there are no tags applied (not sure if that's enforced somewhere else), so that we don't get an "index out of range" runtime error when accessing `thread.applied_tags[0]`

bagelface reviewed (approved):
lgtm
